cpp/src/Ice/IdleTimeoutTransceiverDecorator.h

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#ifndef ICE_IDLE_TIMEOUT_TRANSCEIVER_DECORATOR_H
+#define ICE_IDLE_TIMEOUT_TRANSCEIVER_DECORATOR_H
+
+#include "IceUtil/Timer.h"
+#include "Transceiver.h"
+
+namespace IceInternal
+{
+    // Decorates Transceiver to send heartbeats and detect when no byte is received/read for a while.
+    // This decorator must not be applied on UDP "connections".
+    class IdleTimeoutTransceiverDecorator final : public Transceiver
+    {
+    public:
+        IdleTimeoutTransceiverDecorator(
+            const TransceiverPtr& decoratee,
+            std::chrono::milliseconds readIdleTimeout,
+            std::chrono::milliseconds writeIdleTimeout,
+            const IceUtil::TimerPtr& timer)
+            : _decoratee(decoratee),
+              _readIdleTimeout(readIdleTimeout),
+              _writeIdleTimeout(writeIdleTimeout),
+              _timer(timer)
+        {
+        }
+
+        ~IdleTimeoutTransceiverDecorator();
```

Comment:
Note that Transceiver doesn't have a virtual dtor (or any dtor). Doesn't matter here but that goes against the usual C++ pattern.